gobblin-core/src/main/java/gobblin/source/DatePartitionedSource.java

```diff
@@ -0,0 +1,279 @@
+    this.fsHelper.connect();
+  }
+
+  private void init(SourceState state) {
+    DateTimeZone.setDefault(DateTimeZone.forID("America/Los_Angeles"));
+    try {
+      initFileSystemHelper(state);
+    } catch (FileBasedHelperException e) {
+      Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+    init(state);
+
+    // Initialize filesystem variables
+    HadoopFsHelper fsHelper = (HadoopFsHelper) this.fsHelper;
+    this.fs = fsHelper.getFileSystem();
+
+    // Get the starting date
+    LOG.info("Retrieving the low water mark based on the previous execution");
+
+    DateTime lowWaterMark = getLowWaterMark(state);
+
+    LOG.info("The low water mark for this run is " + lowWaterMark);
+
+    // Determine how many files to process
+    int maxFilesPerJob =
+        state
+            .getPropAsInt(DATE_PARTITIONED_SOURCE_MAX_FILES_PER_JOB, DEFAULT_DATE_PARTITIONED_SOURCE_MAX_FILES_PER_JOB);
+
+    int maxWorkUnitsPerJob =
+        state.getPropAsInt(DATE_PARTITIONED_SOURCE_MAX_WORKUNITS_PER_JOB,
+            DEFAULT_DATE_PARTITIONED_SOURCE_MAX_WORKUNITS_PER_JOB);
+
+    LOG.info("Will pull data from " + lowWaterMark + " until " + maxFilesPerJob + " files have been processed");
+
+    // Create list of workunits
+    LOG.info("Creating workunits");
+
+    // Create namespace, tablename, and table type for the workunits
+    TableType tableType = TableType.valueOf(state.getProp(ConfigurationKeys.EXTRACT_TABLE_TYPE_KEY).toUpperCase());
+
+    // Weighted Queue that all the workunits will be added to
+    MultiWorkUnitWeightedQueue multiWorkUnitWeightedQueue = new MultiWorkUnitWeightedQueue(maxWorkUnitsPerJob);
+
+    // Add failed work units from the previous execution
+    int fileCount = addPreviousWorkUnits(state, multiWorkUnitWeightedQueue);
+
+    if (fileCount >= maxFilesPerJob) {
+      LOG.info("The number of work units from previous job has already reached the upper limit, no more workunits will be made");
+      return multiWorkUnitWeightedQueue.getList();
+    }
+
+    // Process all data from the lowWaterMark date until the maxFilesPerJob has been hit
+    addNewWorkUnits(state, lowWaterMark, fileCount, maxFilesPerJob, tableType, multiWorkUnitWeightedQueue);
+
+    return multiWorkUnitWeightedQueue.getList();
+  }
+
+  /**
+   * Helper method to process the failed {@link WorkUnit} from the previous run and add them to the multiWorkUnitWeightedQueue
+   * @param state
+   * @param multiWorkUnitWeightedQueue
+   * @return
+   */
+  private int addPreviousWorkUnits(SourceState state, MultiWorkUnitWeightedQueue multiWorkUnitWeightedQueue) {
+    int fileCount = 0;
+
+    // Add all workunits that failed in the previous run
+    for (WorkUnit wu : this.getPreviousWorkUnitsForRetry(state)) {
+
+      try {
+        multiWorkUnitWeightedQueue.addWorkUnit(wu,
+            this.fs.getFileStatus(new Path(wu.getProp(ConfigurationKeys.SOURCE_FILEBASED_FILES_TO_PULL))).getLen());
```

Comment:
I moved a majority of them to the init() method. For this specific case, it is iterating over a list of WorkUnits and getting the config from each WorkUnit, so I left it here.